src/rgw/rgw_datalog.h

```diff
@@ -125,6 +125,40 @@ struct RGWDataChangesLogMarker {
   RGWDataChangesLogMarker() = default;
 };
 
+struct rgw_data_notify_entry {
+  std::string key;
+  uint64_t gen = 0;
+
+  rgw_data_notify_entry(const std::string key, uint64_t gen): key(key), gen(gen) {}
+
+  rgw_data_notify_entry(const std::string& strKey = "")
+  : key(strKey) {}
+
+  void encode(ceph::buffer::list& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(key, bl);
+    encode(gen, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(ceph::buffer::list::const_iterator& bl) {
+     DECODE_START(1, bl);
+     decode(key, bl);
+     decode(gen, bl);
+     DECODE_FINISH(bl);
+  }
+
+  void dump(ceph::Formatter* f) const;
+  void decode_json(JSONObj* obj);
+
+  rgw_data_notify_entry& operator=(const rgw_data_notify_entry&) = default;
+
+  bool operator<(const rgw_data_notify_entry& d) const {
+    return (key < d.key && gen < d.gen);
+  }
+};
+WRITE_CLASS_ENCODER(rgw_data_notify_entry)
```

Comment:
this struct is only encoded in json form for the notify2 API. we won't ever need to binary encode it, so you can skip this and encode()/decode()